Spigot-Server-Patches/0003-MC-Dev-fixes.patch

```diff
@@ -114,8 +114,108 @@ index 58f47d0d..8860a012 100644
          }
  
          this.b.set(i, t0);
+diff --git a/src/main/java/net/minecraft/server/ServerPing.java b/src/main/java/net/minecraft/server/ServerPing.java
+index 2179664a0..10bc7fc1f 100644
+--- a/src/main/java/net/minecraft/server/ServerPing.java
++++ b/src/main/java/net/minecraft/server/ServerPing.java
+@@ -57,7 +57,7 @@ public class ServerPing {
+ 
+         public Serializer() {}
+ 
+-        public ServerPing a(JsonElement jsonelement, Type type, JsonDeserializationContext jsondeserializationcontext) throws JsonParseException {
```

Comment:
Why was this done? If anything uses the obfuscated names, it would now throw. Silly if they did yes, but why add a diff to maintain?

We should just add obfuscation helpers here if needed.

Reply:
This is just a compile fix because Spigot's deofuscator doesn't handle generic interfaces properly. The methods with the real name (`serialize` and `deserialize`) are the type erased bridge methods that shouldn't appear in the decompiled source code at all. The actual methods implementing the interface remained obfuscated.

Without this patch, it will throw a compile error because it doesn't implement the interface properly:

```java
Error:(56, 19) java: net.minecraft.server.ServerPing.Serializer is not abstract and does not override abstract method deserialize(com.google.gson.JsonElement,java.lang.reflect.Type,com.google.gson.JsonDeserializationContext) in com.google.gson.JsonDeserializer
Error:(109, 23) java: deserialize(com.google.gson.JsonElement,java.lang.reflect.Type,com.google.gson.JsonDeserializationContext) in net.minecraft.server.ServerPing.Serializer cannot implement deserialize(com.google.gson.JsonElement,java.lang.reflect.Type,com.google.gson.JsonDeserializationContext) in com.google.gson.JsonDeserializer
  return type java.lang.Object is not compatible with net.minecraft.server.ServerPing
Error:(105, 28) java: name clash: serialize(java.lang.Object,java.lang.reflect.Type,com.google.gson.JsonSerializationContext) in net.minecraft.server.ServerPing.Serializer and serialize(T,java.lang.reflect.Type,com.google.gson.JsonSerializationContext) in com.google.gson.JsonSerializer have the same erasure, yet neither overrides the other
Error:(132, 23) java: net.minecraft.server.ServerPing.ServerData.Serializer is not abstract and does not override abstract method deserialize(com.google.gson.JsonElement,java.lang.reflect.Type,com.google.gson.JsonDeserializationContext) in com.google.gson.JsonDeserializer
Error:(154, 27) java: deserialize(com.google.gson.JsonElement,java.lang.reflect.Type,com.google.gson.JsonDeserializationContext) in net.minecraft.server.ServerPing.ServerData.Serializer cannot implement deserialize(com.google.gson.JsonElement,java.lang.reflect.Type,com.google.gson.JsonDeserializationContext) in com.google.gson.JsonDeserializer
  return type java.lang.Object is not compatible with net.minecraft.server.ServerPing.ServerData
Error:(150, 32) java: name clash: serialize(java.lang.Object,java.lang.reflect.Type,com.google.gson.JsonSerializationContext) in net.minecraft.server.ServerPing.ServerData.Serializer and serialize(T,java.lang.reflect.Type,com.google.gson.JsonSerializationContext) in com.google.gson.JsonSerializer have the same erasure, yet neither overrides the other
Error:(187, 23) java: net.minecraft.server.ServerPing.ServerPingPlayerSample.Serializer is not abstract and does not override abstract method deserialize(com.google.gson.JsonElement,java.lang.reflect.Type,com.google.gson.JsonDeserializationContext) in com.google.gson.JsonDeserializer
Error:(242, 27) java: deserialize(com.google.gson.JsonElement,java.lang.reflect.Type,com.google.gson.JsonDeserializationContext) in net.minecraft.server.ServerPing.ServerPingPlayerSample.Serializer cannot implement deserialize(com.google.gson.JsonElement,java.lang.reflect.Type,com.google.gson.JsonDeserializationContext) in com.google.gson.JsonDeserializer
  return type java.lang.Object is not compatible with net.minecraft.server.ServerPing.ServerPingPlayerSample
Error:(238, 32) java: name clash: serialize(java.lang.Object,java.lang.reflect.Type,com.google.gson.JsonSerializationContext) in net.minecraft.server.ServerPing.ServerPingPlayerSample.Serializer and serialize(T,java.lang.reflect.Type,com.google.gson.JsonSerializationContext) in com.google.gson.JsonSerializer have the same erasure, yet neither overrides the other
```

Reply:
Ok, good with me then.

Reply:
Could you please mark these as a decompiler fix in the patch, makes it easier to understand what is going on further down the line
Rebase, and it looks fine to me!